api/v1beta1/jujucluster_types.go

```diff
@@ -36,7 +40,141 @@ type APIEndpoint struct {
+	HostCloudRegion string `json:"hostCloudRegion,omitempty"`
+
+	// Description describes the type of cloud.
+	Description string `json:"description,omitempty"`
+
+	// AuthTypes are the authentication modes supported by the cloud.
+	AuthTypes AuthTypes `json:"authTypes"`
+
+	// Endpoint is the default endpoint for the cloud regions, may be
+	// overridden by a region.
+	Endpoint string `json:"endpoint,omitempty"`
+
+	// IdentityEndpoint is the default identity endpoint for the cloud
+	// regions, may be overridden by a region.
+	IdentityEndpoint string `json:"identityEndpoint,omitempty"`
+
+	// StorageEndpoint is the default storage endpoint for the cloud
+	// regions, may be overridden by a region.
+	StorageEndpoint string `json:"storageEndpoint,omitempty"`
+
+	// Regions are the regions available in the cloud.
+	//
+	// Regions is a slice, and not a map, because order is important.
+	// The first region in the slice is the default region for the
+	// cloud.
+	Regions []Region `json:"regions,omitempty"`
+
+	// Config contains optional cloud-specific configuration to use
+	// when bootstrapping Juju in this cloud. The cloud configuration
+	// will be combined with Juju-generated, and user-supplied values;
+	// user-supplied values taking precedence.
+	// This should be a map[string]interface{}, but kube-builder does not allow interface{} types
+	// so we are forced to use untyped JSON here
+	Config *apiextensionsv1.JSON `json:"config,omitempty"`
+
+	// RegionConfig contains optional region specific configuration.
+	// Like Config above, this will be combined with Juju-generated and user
+	// supplied values; with user supplied values taking precedence.
+	// This should be a map[string]map[string]interface{}, but kube-builder does not allow interface{} types
+	// so we are forced to use untyped JSON here
+	RegionConfig *apiextensionsv1.JSON `json:"regionConfig,omitempty"`
+
+	// CACertificates contains an optional list of Certificate
+	// Authority certificates to be used to validate certificates
+	// of cloud infrastructure components
+	// The contents are Base64 encoded x.509 certs.
+	CACertificates []string `json:"CACertificates,omitempty"`
+
+	// SkipTLSVerify is true if the client should be asked not to
+	// validate certificates. It is not recommended for production clouds.
+	// It is secure (false) by default.
+	SkipTLSVerify bool `json:"skipTLSVerify,omitempty"`
+
+	// IsControllerCloud is true when this is the cloud used by the controller.
+	IsControllerCloud bool `json:"isControllerCloud,omitempty"`
+}
+
+type Credential struct {
+	// CredentialSecretName is used to tell the controller the name of the secret containing the cloud credentials
+	CredentialSecretName string `json:"credentialSecretName"`
+
+	// CredentialSecretNamespace is used to tell the controller the namespace of the secret containing the cloud credentials
+	CredentialSecretNamespace string `json:"credentialSecretNamespace"`
+}
+
```

Comment:
Since constraints.value is not in the api package, controller-gen cant generate the deepcopy methods like it usually does and we have to add our own implementation